### ISSUE TITLE
Cannot Start Grafana After Ubuntu Upgrade from 18 to 20

### ISSUE THREAD

**Issue by pillbug22** [grafana/grafana#30755]
**What happened**:
After upgrading from Ubuntu 18.04 to Ubuntu 20.04, grafana does not start (used on server with Icinag2 montoring server). After upgrading this box from Ubuntu 16 to 18 in the past, the service would not auto-start on bootup, but we could manually start it after boot with
`systemctl start grafana-server`

Attempting to start grafana manually now also errors with
```
~$ sudo systemctl start grafana-server
Job for grafana.service failed because the service did not take the steps required by its unit configuration.
See "systemctl status grafana.service" and "journalctl -xe" for details.
```

systemctl status grafana-server output:
```
~$ sudo systemctl status grafana-server
● grafana.service - LSB: Start grafana at boot time
     Loaded: loaded (/etc/init.d/grafana; generated)
     Active: failed (Result: protocol) since Thu 2020-12-31 11:28:06 CST; 1s ago
       Docs: man:systemd-sysv-generator(8)
    Process: 18292 ExecStart=/etc/init.d/grafana start (code=exited, status=0/SUCCESS)

Dec 31 11:28:06 hostname systemd[1]: Starting LSB: Start grafana at boot time...
Dec 31 11:28:06 hostname systemd[1]: grafana.service: Can't open PID file /run/grafana.pid (yet?) after start: Operation not permitted
Dec 31 11:28:06 hostname systemd[1]: grafana.service: Failed with result 'protocol'.
Dec 31 11:28:06 hostname systemd[1]: Failed to start LSB: Start grafana at boot time.
```

Output from journalctl -xe
```
Dec 31 11:29:59 hostname sudo[21006]: admin : TTY=pts/0 ; PWD=/home/admin ; USER=root ; COMMAND=/bin/systemctl start grafana-server
Dec 31 11:29:59 hostname sudo[21006]: pam_unix(sudo:session): session opened for user root by admin(uid=0)
Dec 31 11:29:59 hostname systemd[1]: Starting LSB: Start grafana at boot time...
-- Subject: A start job for unit grafana.service has begun execution
-- Defined-By: systemd
-- Support: http://www.ubuntu.com/support
--
-- A start job for unit grafana.service has begun execution.
--
-- The job identifier is 3052.
Dec 31 11:29:59 hostname systemd[1]: grafana.service: Can't open PID file /run/grafana.pid (yet?) after start: Operation not permitted
Dec 31 11:29:59 hostname sudo[21006]: pam_unix(sudo:session): session closed for user root
Dec 31 11:29:59 hostname systemd[1]: grafana.service: Failed with result 'protocol'.
-- Subject: Unit failed
-- Defined-By: systemd
-- Support: http://www.ubuntu.com/support
--
-- The unit grafana.service has entered the 'failed' state with result 'protocol'.
Dec 31 11:29:59 hostname systemd[1]: Failed to start LSB: Start grafana at boot time.
-- Subject: A start job for unit grafana.service has failed
-- Defined-By: systemd
-- Support: http://www.ubuntu.com/support
--
-- A start job for unit grafana.service has finished with a failure.
--
-- The job identifier is 3052 and the job result is failed.
```

The last entries in /var/log/grafana/grafana.log are from the system shutting down during the Ubuntu upgrade.

- deb https://packages.grafana.com/enterprise/deb stable main is included in
/etc/apt/sources.list.d/grafana.list
- sudo apt update and sudo apt upgrade grafana state it is current versions
- checking current package version says is current:
:~$ sudo apt-cache policy grafana
grafana:
Installed: 7.3.7
Candidate: 7.3.7

Checking config at /etc/grafana/grafana.ini , it appears this file is still in place because it does have items we had set such as hostname. I found [1 other report for the ‘protocol’ error](https://github.com/grafana/grafana/issues/24464), but it was closed with no resolution. Also [posted on the Community forums](https://community.grafana.com/t/grafana-server-not-starting-after-upgrade-to-ubuntu-20/41183), but no responses. Any thoughts on what to check next?

**What you expected to happen**:
grafana to start

**How to reproduce it (as minimally and precisely as possible)**:
`systemctl start grafana-server`

**Anything else we need to know?**:

**Environment**:
- Grafana version: 7.3.7
- Data source type & version: Icinga2 via InfluxDB
- OS Grafana is installed on: Ubuntu 20.04
- User OS & Browser: Win10 20H2 / Chrome 87
- Grafana plugins: None aware
- Others:

**Comment by torkelo** [comment]
Strange, some kind of permission issue? 

Can you check what your /etc/default/grafana file says?

Should be a line there with 
```
# Only used on systemd systems
PID_FILE_DIR=/var/run/grafana
```

One opion to try is to uninstall & reinstalll (with proper backups in place)

**Comment by pillbug22** [comment]
Contents of /etc/default/grafana:


```
GRAFANA_USER=grafana

GRAFANA_GROUP=grafana

GRAFANA_HOME=/usr/share/grafana

LOG_DIR=/var/log/grafana

DATA_DIR=/var/lib/grafana

MAX_OPEN_FILES=10000

CONF_DIR=/etc/grafana

CONF_FILE=/etc/grafana/grafana.ini

RESTART_ON_UPGRADE=false
```

**Comment by torkelo** [comment]
Added this line to that file 

```
PID_FILE_DIR=/var/run/grafana
```

**Comment by pillbug22** [comment]
Added, rebooted for good measure.

After restart, systemctl status grafana-server:

```
● grafana.service - LSB: Start grafana at boot time
     Loaded: loaded (/etc/init.d/grafana; generated)
     Active: failed (Result: protocol) since Tue 2021-02-02 07:54:49 CST; 1min 48s ago
       Docs: man:systemd-sysv-generator(8)
    Process: 961 ExecStart=/etc/init.d/grafana start (code=exited, status=0/SUCCESS)

Feb 02 07:54:48 hostname systemd[1]: Starting LSB: Start grafana at boot time...
Feb 02 07:54:49 hostname systemd[1]: grafana.service: Can't open PID file /run/grafana.pid (yet?) after start: Operation not permitted
Feb 02 07:54:49 hostname systemd[1]: grafana.service: Failed with result 'protocol'.
Feb 02 07:54:49 hostname systemd[1]: Failed to start LSB: Start grafana at boot time.
```

systemctl start grafana-server
```
Job for grafana.service failed because the service did not take the steps required by its unit configuration.
See "systemctl status grafana.service" and "journalctl -xe" for details.

```



systemctl status grafana-server:

```
● grafana.service - LSB: Start grafana at boot time
     Loaded: loaded (/etc/init.d/grafana; generated)
     Active: failed (Result: protocol) since Tue 2021-02-02 07:56:51 CST; 2s ago
       Docs: man:systemd-sysv-generator(8)
    Process: 4569 ExecStart=/etc/init.d/grafana start (code=exited, status=0/SUCCESS)

Feb 02 07:56:51 hostname systemd[1]: Starting LSB: Start grafana at boot time...
Feb 02 07:56:51 hostname systemd[1]: grafana.service: Can't open PID file /run/grafana.pid (yet?) after start: Operation not permitted
Feb 02 07:56:51 hostname systemd[1]: grafana.service: Failed with result 'protocol'.
Feb 02 07:56:51 hostname systemd[1]: Failed to start LSB: Start grafana at boot time.

```
journalctl -xe

```
Feb 02 07:56:37 hostname sudo[3962]: lcsadmin : TTY=pts/0 ; PWD=/home/lcsadmin ; USER=root ; COMMAND=/bin/systemctl status grafana-server
Feb 02 07:56:37 hostname sudo[3962]: pam_unix(sudo:session): session opened for user root by lcsadmin(uid=0)
Feb 02 07:56:41 hostname sudo[3962]: pam_unix(sudo:session): session closed for user root
Feb 02 07:56:45 hostnamet influxd[947]: [httpd] 127.0.0.1 - icinga2_grafana_user [02/Feb/2021:07:56:45 -0600] "POST /write?db=icinga2_grafana&p=%5BREDACTED%5D&precision=s&u=ic>
Feb 02 07:56:51 hostname audit[817]: AVC apparmor="ALLOWED" operation="open" profile="/usr/sbin/sssd//null-/usr/libexec/sssd/sssd_nss" name="/proc/4566/cmdline" pid=817 comm=>
Feb 02 07:56:51 hostname kernel: audit: type=1400 audit(1612274211.216:2274): apparmor="ALLOWED" operation="open" profile="/usr/sbin/sssd//null-/usr/libexec/sssd/sssd_nss" na>
Feb 02 07:56:51 hostname sudo[4566]: lcsadmin : TTY=pts/0 ; PWD=/home/lcsadmin ; USER=root ; COMMAND=/bin/systemctl start grafana-server
Feb 02 07:56:51 hostname sudo[4566]: pam_unix(sudo:session): session opened for user root by lcsadmin(uid=0)
Feb 02 07:56:51 hostname systemd[1]: Starting LSB: Start grafana at boot time...
-- Subject: A start job for unit grafana.service has begun execution
-- Defined-By: systemd
-- Support: http://www.ubuntu.com/support
--
-- A start job for unit grafana.service has begun execution.
--
-- The job identifier is 1092.
Feb 02 07:56:51 hostname systemd[1]: grafana.service: Can't open PID file /run/grafana.pid (yet?) after start: Operation not permitted
Feb 02 07:56:51 hostname sudo[4566]: pam_unix(sudo:session): session closed for user root
Feb 02 07:56:51 hostname systemd[1]: grafana.service: Failed with result 'protocol'.
-- Subject: Unit failed
-- Defined-By: systemd
-- Support: http://www.ubuntu.com/support
--
-- The unit grafana.service has entered the 'failed' state with result 'protocol'.
Feb 02 07:56:51 hostname systemd[1]: Failed to start LSB: Start grafana at boot time.
-- Subject: A start job for unit grafana.service has failed
-- Defined-By: systemd
-- Support: http://www.ubuntu.com/support
--
-- A start job for unit grafana.service has finished with a failure.
--
-- The job identifier is 1092 and the job result is failed.

```

**Comment by pillbug22** [comment]
- at remove grafana
- Re-installed (Enterprise) per https://grafana.com/docs/grafana/latest/installation/debian
- systemctl daemon-reload
- systemctl start grafana-server
- receiving same error that could not start due to protocol error

**Comment by torkelo** [comment]
I just spun up a ubuntu 20 LTS and installed grafana enterprise using the deb repo, was not able to replicate your PID write issue  

Can you check that the /var/run/grafana directory exists and the grafana user has permissions on this directory ? 
 

```
torkel@ubuntu-20:/var/run$ ls -la | grep grafana
drwxr-x---  2 grafana grafana   60 Feb  9 15:17 grafana
```

**Comment by pillbug22** [comment]
- grafana directory wasn't showing under /var/run
```
/var/run$ sudo mkdir /var/run/grafana
/var/run$ sudo chown grafana:grafana /var/run/grafana
/var/run$ ls -la | grep grafana
drwxr-xr-x  2 grafana  grafana    40 Feb  9 11:22 grafana
/var/run$ sudo systemctl start grafana-server
Job for grafana.service failed because the service did not take the steps required by its unit configuration.
See "systemctl status grafana.service" and "journalctl -xe" for details.
```
- received same protocol / grafana.pid error in journalctl

**Comment by ivanahuckova** [comment]
@torkelo do we need any additional info for this?

**Comment by torkelo** [comment]
Closing this as we have no other reports like this and failed to replicate, somehow the /var/run/grafana was not created during grafana install, try to manually add it and set chown to grafana user

**Comment by sandipb** [comment]
@torkelo 

I am seeing this issue in a new install.

From the deb source, I don't see `/var/run/grafana` being created in https://github.com/grafana/grafana/blob/main/packaging/deb/control/postinst

Could that be the reason?

**Comment by ivanahuckova** [comment]
+cc @grafana/backend-platform

**Comment by blackw1ng** [comment]
@torkelo also see this exact behavior on Ubuntu 20.04 LTS after upgrading from 18.04 LTS.

```
Sep 13 22:38:50 systemd[1]: Starting LSB: Start grafana at boot time...
-- Subject: A start job for unit grafana.service has begun execution
-- Defined-By: systemd
-- Support: http://www.ubuntu.com/support
-- 
-- A start job for unit grafana.service has begun execution.
-- 
-- The job identifier is 3693.
Sep 13 22:38:50 systemd[1]: grafana.service: Can't open PID file /run/grafana.pid (yet?) after start: Operation not permitted
Sep 13 22:38:50 systemd[1]: grafana.service: Failed with result 'protocol'.
```

I manually created `/var/run/grafana` (and chowned) and added the `PID_FILE_DIR=/var/run/grafana` to the defaults.

**Comment by pillbug22** [comment]
> I manually created `/var/run/grafana` (and chowned) and added the `PID_FILE_DIR=/var/run/grafana` to the defaults.

@blackw1ng did you add `PID_FILE_DIR=/var/run/grafana` to /etc/environment ? I checked that my /var/run/grafana has the permissions as @torkelo suggested earlier, but there wasn't any entry for PID_FILE_DIR in the environment variables file. Added it there, rebooted, still save error when trying to start up the service (failed with result 'protocol').

Interestingly, the error message does still say Can't open PID file /run/grafana.pid , rather than I would expect it to now say /var/run/grafana.pid 

Maybe need to manually create the grafana folder directly under /run instead of /var/run ?

**Comment by blackw1ng** [comment]
So - i checked all that... 
- [x] the `/var/run` is indeed a symlink to `/run` (so it doesn't matter)
- [x] the `/etc/init.d/grafana` defines a `PID_FILE=/var/run/grafana/$NAME.pid`... however this someone does not take place and is not applied.
- [x] permissions of `/run/grafana/` are `rwx` to `grafana:grafana`

something on the interaction of `systemd` and the legacy sysVinit script seems to be... off. 
The message that strikes me is the "(yet?)" in the message...
```
systemd[1]: grafana.service: Can't open PID file /run/grafana.pid (yet?) after start: Operation not permitted
systemd[1]: grafana.service: Failed with result 'protocol'.
```

**Comment by blackw1ng** [comment]
@pillbug22 so, here is what I did so far, that solved the issue
TL;DR: the `grafana` service is left over, the right one is `grafana-server` - and you need to purge the old one.

(all as root / with sudo)
1. `apt purge grafana`
2. `systemctl disable grafana`
3. `apt install grafana`
4. `systemctl enable grafana-server`
5. `mkdir /var/run/grafana` (seems to be missing in install script)
6. `chown grafana:grafana /var/run/grafana`
7. `systemctl start grafana-server`

then i get
```
# service grafana-server status
● grafana-server.service - Grafana instance
     Loaded: loaded (/usr/lib/systemd/system/grafana-server.service; enabled; vendor preset: enabled)
     Active: active (running) since Wed 2021-09-15 16:46:01 CEST; 32s ago
```

🎉

**Comment by pillbug22** [comment]
Followed along your steps, then got:

```
~$ sudo systemctl start grafana-server
~$ sudo systemctl status grafana-server
● grafana-server.service - Grafana instance
     Loaded: loaded (/usr/lib/systemd/system/grafana-server.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Wed 2021-09-15 10:24:02 CDT; 2s ago
       Docs: http://docs.grafana.org
    Process: 3609968 ExecStart=/usr/sbin/grafana-server --config=${CONF_FILE} --pidfile=${PID_FILE_DIR}/grafana-server.pid --packaging=deb cfg:default.paths.logs=${LOG_DIR} cfg:default.paths.data=>
   Main PID: 3609968 (code=exited, status=1/FAILURE)

Sep 15 10:24:02 systemd[1]: grafana-server.service: Scheduled restart job, restart counter is at 5.
Sep 15 10:24:02  systemd[1]: Stopped Grafana instance.
Sep 15 10:24:02  systemd[1]: grafana-server.service: Start request repeated too quickly.
Sep 15 10:24:02  systemd[1]: grafana-server.service: Failed with result 'exit-code'.
Sep 15 10:24:02  systemd[1]: Failed to start Grafana instance.
```

Checking /var/log/grafana/grafana.log  : 

```
t=2021-09-15T10:24:01-0500 lvl=info msg="Starting Grafana" logger=server version=8.1.3 commit=a61f38238c branch=HEAD compiled=2021-09-08T04:32:03-0500
t=2021-09-15T10:24:01-0500 lvl=info msg="Config loaded from" logger=settings file=/usr/share/grafana/conf/defaults.ini
t=2021-09-15T10:24:01-0500 lvl=info msg="Config loaded from" logger=settings file=/etc/grafana/grafana.ini
t=2021-09-15T10:24:01-0500 lvl=info msg="Config overridden from command line" logger=settings arg="default.paths.data=/var/lib/grafana"
t=2021-09-15T10:24:01-0500 lvl=info msg="Config overridden from command line" logger=settings arg="default.paths.logs=/var/log/grafana"
t=2021-09-15T10:24:01-0500 lvl=info msg="Config overridden from command line" logger=settings arg="default.paths.plugins=/var/lib/grafana/plugins"
t=2021-09-15T10:24:01-0500 lvl=info msg="Config overridden from command line" logger=settings arg="default.paths.provisioning=/etc/grafana/provisioning"
t=2021-09-15T10:24:01-0500 lvl=info msg="Path Home" logger=settings path=/usr/share/grafana
t=2021-09-15T10:24:01-0500 lvl=info msg="Path Data" logger=settings path=/var/lib/grafana
t=2021-09-15T10:24:01-0500 lvl=info msg="Path Logs" logger=settings path=/var/log/grafana
t=2021-09-15T10:24:01-0500 lvl=info msg="Path Plugins" logger=settings path=/var/lib/grafana/plugins
t=2021-09-15T10:24:01-0500 lvl=info msg="Path Provisioning" logger=settings path=/etc/grafana/provisioning
t=2021-09-15T10:24:01-0500 lvl=info msg="App mode production" logger=settings
t=2021-09-15T10:24:01-0500 lvl=info msg="Writing PID file" logger=server path=/var/run/grafana/grafana-server.pid pid=3609968
t=2021-09-15T10:24:01-0500 lvl=info msg="Connecting to DB" logger=sqlstore dbtype=sqlite3
t=2021-09-15T10:24:01-0500 lvl=info msg="Starting DB migrations" logger=migrator
t=2021-09-15T10:24:01-0500 lvl=info msg="migrations completed" logger=migrator performed=0 skipped=385 duration=1.55151ms
t=2021-09-15T10:24:01-0500 lvl=info msg="Validated license token" logger=licensing appURL=http://<internal FQDN>:3000/ source=disk status=NotFound
t=2021-09-15T10:24:01-0500 lvl=info msg="Starting plugin search" logger=plugins
t=2021-09-15T10:24:01-0500 lvl=warn msg="Skipping plugin as it's a duplicate" logger=plugins id=stackdriver
t=2021-09-15T10:24:01-0500 lvl=warn msg="Some plugin scanning errors were found" logger=plugins errors="plugin with ID 'stackdriver' already exists in '/usr/share/grafana/public/app/plugins/datasource/stackdriver'"
t=2021-09-15T10:24:01-0500 lvl=warn msg="The following files were not included in the signature" logger=plugins plugin=input files=[LICENSE]
```

The only thing that partially sticks out is that it can't find a license, which we've never had on it, rather just using the download from grafana.com . So that shouldn't be an issue, it's just verifying it's not a paid install to enable the full enterprise features. 

Ah. Checking journalctl for grafana-server, I see 1 extra line that isn't included in the grafana.log :

```Sep 15 10:39:17  grafana-server[3633233]: service init failed: failed to scan bundled plugins directory '/usr/share/grafana/plugins-bundled': unknown plugin type "transform"```

Taking a look under /usr/share/grafana/plugins-bundled/ , there is a sub-folder "gel-0.6.0" , which checking the README, says that plugin is an alpha-stage plugin for backend transformations. Made a backup of that folder, removed it, re-ran systemctl start grafana-server and BOOM! it's up and running.

Now we have to figure out an issue where the section in Icinga web where it should be showing the graphs from Grafana instead shows [an error](https://community.icinga.com/uploads/default/original/2X/f/f00aee2d375931ce9bf439150a865d2d3293af77.png),  but at least Grafana itself is back up and running (and already can find some discussions about the same error). We can login to the Grafana URL and view any graphs/dashboards OK.

Thanks to @blackw1ng for a push in the right direction and also a head's up that new plugin is causing issues for at least 1 install

**Comment by quantumzervices** [comment]
I just experienced this issue when I upgraded from 8.1.2 to 8.2.0-beta1.  I tried a host of troubleshooting but I had to revert back to the previous version by using 'apt install grafana=8.1.2'.
